src/Castle.Windsor/Windsor/WindsorContainer.cs

```diff
@@ -541,7 +541,15 @@ public virtual object Resolve(Type service, IDictionary arguments)
 		/// <returns></returns>
 		public virtual object Resolve(Type service, object argumentsAsAnonymousType)
 		{
-			return Resolve(service, new ReflectionBasedDictionaryAdapter(argumentsAsAnonymousType));
+			if (argumentsAsAnonymousType is IReadOnlyDictionary<string, object>)
+			{
+				return kernel.Resolve(service, CloneDictionary((IReadOnlyDictionary<string, object>) argumentsAsAnonymousType));
+			}
+			if (argumentsAsAnonymousType is IDictionary)
+			{
+				return kernel.Resolve(service, (IDictionary)argumentsAsAnonymousType);
+			}
+			return kernel.Resolve(service, new ReflectionBasedDictionaryAdapter(argumentsAsAnonymousType));
 		}
```

Comment:
@jonorossi 

Example of type sniffing acrobatics. Also appears to be a recurring theme.